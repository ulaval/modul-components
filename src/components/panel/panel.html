<article :class="[ componentName,
                {'is-primary': type!='secondary',
                'is-secondary': type=='secondary',
                'no-shadow': hasShadow == 'false',
                'no-border': hasBorder == 'false',
                'no-padding': hasPadding == 'false',
                'no-padding-header': hasPaddingHeader == 'false',
                'no-padding-body': hasPaddingBody == 'false'} ]">
    <header :class="[ componentName + '__header']" v-if="hasHeader">
        <slot name="header"></slot>
        <span :class="[ componentName + '__header__content', 'is-left']" v-if="hasContentLeft">
            <slot name="header-content-left"></slot>
        </span>
        <span :class="[ componentName + '__header__content', 'is-right']" v-if="hasContentRight">
            <slot name="header-content-right"></slot>
        </span>
    </header>
    <div :class="[ componentName + '__body']">
        <slot>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas iaculis nibh eu diam pellentesque lobortis. Duis mattis vel orci at luctus. Fusce vitae eros quis diam iaculis imperdiet quis vitae sapien. Aliquam blandit commodo arcu et vehicula. Aliquam erat volutpat. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Donec vehicula condimentum elementum. Sed laoreet est eget lorem tristique condimentum. Vestibulum at odio orci. Duis nisl erat, aliquam non tortor et, dapibus maximus nulla. Mauris lobortis, orci nec lobortis faucibus, nisi justo eleifend erat, sed convallis velit mi non lacus. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas.</slot>
    </div>
</article>