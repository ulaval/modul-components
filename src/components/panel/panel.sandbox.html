<div>
    <m-panel>
        <h2 slot="header">I am a title</h2>
        <p>Lorem ipsum dolor, sit amet consectetur adipisicing elit. Earum voluptas quis corporis fugit nobis beatae vel, quia at consequuntur odit ut excepturi omnis ea explicabo blanditiis quaerat cupiditate odio a?</p>
    </m-panel>
</div>
