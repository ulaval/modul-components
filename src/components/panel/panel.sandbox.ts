import Vue, { PluginObject } from 'vue';
import { Component } from 'vue-property-decorator';

import { PANEL_NAME } from '../component-names';
import WithRender from './panel.sandbox.html';

@WithRender
@Component
export class MPanelSandbox extends Vue {
}

const PanelSandboxPlugin: PluginObject<any> = {
    install(v, options): void {
        v.component(`${PANEL_NAME}-sandbox`, MPanelSandbox);
    }
};

export default PanelSandboxPlugin;
