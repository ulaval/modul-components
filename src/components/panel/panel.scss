@import 'abstracts/variables';
$m-panel--screen--max: 639px;
$m-panel--screen--min: 640px;

.m-panel {
    position: relative;
    display: flex;
    flex-direction: column;

    @include m-is-ie() {
        display: block;
    }

    &.m--is-skin-light {
        background: $m-color--white;
    }

    &.m--is-skin-dark {
        background: $m-color--grey-light;
    }

    &.m--is-skin-darker {
        background: $m-color--grey-darker;
        color: $m-color--white;
    }

    &.m--has-border {
        border: $m-border-width--s solid $m-color--border;
    }

    &.m--has-border-large {
        border: $m-border-width--l solid $m-color--border;
    }

    &.m--has-shadow {
        box-shadow: $m-box-shadow--s;
    }

    &.m--is-highlighted {
        border-top: 0;

        &::before {
            position: relative;
            top: 0;
            content: '';
            display: block;
            height: 4px;
            background-color: $m-color--accent;
        }

        &.m--has-border {
            &::before {
                right: - $m-border-width--s;
                left: - $m-border-width--s;
                width: calc(100% + #{$m-border-width--s * 2}); // Magic number
            }
        }

        &.m--has-border-large {
            &::before {
                right: - $m-border-width--l;
                left: - $m-border-width--l;
                width: calc(100% + #{$m-border-width--l * 2}); // Magic number
            }
        }

        &:not(.m--has-border) {
            &::before {
                right: 0;
                left: 0;
                width: 100%;
            }
        }
    }

    &.m--has-padding-large {
        .m-panel {
            &__header,
            &__footer {
                &:not(.m--no-padding) {
                    padding: $m-padding;

                    @media (min-width: $m-mq--min--s) {
                        padding: $m-padding--l;
                    }
                }
            }

            &__body {
                &:not(.m--no-padding) {
                    padding-right: $m-padding;
                    padding-left: $m-padding;

                    @media (min-width: $m-mq--min--s) {
                        padding-left: $m-padding--l;
                        padding-right: $m-padding--l;
                    }
                }

                &.m--has-padding-top {
                    padding-top: $m-padding;

                    @media (min-width: $m-mq--min--s) {
                        padding-top: $m-padding--l;
                    }
                }

                &.m--has-padding-bottom {
                    padding-bottom: $m-padding;

                    @media (min-width: $m-mq--min--s) {
                        padding-bottom: $m-padding--l;
                    }
                }
            }
        }
    }

    &__header,
    &__footer {
        &:not(.m--no-padding) {
            padding: $m-padding--s;

            @media (min-width: $m-mq--min--s) {
                padding: $m-padding;
            }
        }
    }

    &__header {
        position: relative;

        &.m--has-menu {
            padding-right: $m-touch-size;
        }

        &:not(.m--no-padding) {
            &.m--has-menu {
                padding-right: $m-margin--xs + $m-touch-size;
            }
        }

        &__content {
            flex: 1;

            > :first-child {
                margin-top: 0;

                > :first-child {
                    margin-top: 0;
                }
            }
        }

        &__menu {
            .m--is-skin-light & {
                right: $m-spacing;
            }
        }

        &.m--is-right-content-alone {
            &[min-width~="#{ $m-panel--screen--min }"] {
                justify-content: flex-end;
            }
        }
    }

    &__body,
    &__footer {
        > :first-child {
            margin-top: 0;
            padding-top: 0;

            > :first-child {
                margin-top: 0;
            }
        }
    }

    &__body {
        flex: 1 1 auto;
        position: relative;

        &:not(.m--no-padding) {
            padding-right: $m-padding--s;
            padding-left: $m-padding--s;

            @media (min-width: $m-mq--min--s) {
                padding-left: $m-padding;
                padding-right: $m-padding;
            }
        }

        &.m--has-padding-top {
            padding-top: $m-padding;
        }

        &.m--has-padding-bottom {
            padding-bottom: $m-padding;
        }

        @include m-scrollbar();
    }

    &__footer {
        border-top: 1px solid $m-color--grey;
    }

    &[max-width~="#{ $m-panel--screen--max }"] {
        &.m--is-skin-light {
            .m-panel__header {
                display: block;
            }
        }

        &:not(.m--is-right-content-alone) {
            .m-panel__header__content-right {
                margin-top: $m-margin;
            }
        }
    }

    &[min-width~="#{ $m-panel--screen--min }"] {
        .m-panel__header {
            display: flex;
            align-items: center;

            &__content {
                flex: 1 1 auto;
            }
        }

        &:not(.m--is-right-content-alone) {
            .m-panel__header {
                justify-content: space-between;
            }
        }
    }
}
