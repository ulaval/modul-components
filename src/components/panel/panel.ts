import Vue, { PluginObject } from 'vue';
import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';
import WithRender from './panel.html?style=./panel.scss';
import { PANEL_NAME } from '../component-names';
import ElementQueries from 'css-element-queries/src/ElementQueries';

export enum MPanelSkin {
    Light = 'light',
    Dark = 'dark',
    Darker = 'darker'
}

const HEADER_RIGHT_CONTENT: string = 'header-right-content';
const MENU: string = 'menu';

@WithRender
@Component
export class MPanel extends Vue {
    @Prop({
        default: MPanelSkin.Light,
        validator: value =>
            value == MPanelSkin.Light ||
            value == MPanelSkin.Dark ||
            value == MPanelSkin.Darker
    })
    public skin: MPanelSkin;

    @Prop()
    public highlighted: boolean;

    @Prop()
    public shadow: boolean;

    @Prop({ default: true })
    public border: boolean;

    @Prop()
    public borderLarge: boolean;

    @Prop({ default: true })
    public padding: boolean;

    @Prop()
    public paddingLarge: boolean;

    @Prop({ default: true })
    public paddingHeader: boolean;

    @Prop({ default: true })
    public paddingBody: boolean;

    @Prop({ default: true })
    public paddingFooter: boolean;

    protected mounted(): void {
        ElementQueries.init();
    }

    protected beforeDestroy(): void {
        ElementQueries.detach(this.$el);
    }

    private get lightSkin(): boolean {
        return this.skin == MPanelSkin.Light;
    }

    private get darkSkin(): boolean {
        return this.skin == MPanelSkin.Dark;
    }

    private get darkerSkin(): boolean {
        return this.skin == MPanelSkin.Darker;
    }

    private get hasHeader(): boolean {
        if (this.$slots.header || this.$slots[HEADER_RIGHT_CONTENT]) {
            return true;
        }
        return false;
    }

    private get hasHeaderRightContentSlot(): boolean {
        return !!this.$slots[HEADER_RIGHT_CONTENT];
    }

    private get hasHeaderMenuSlot(): boolean {
        return !!this.$slots[MENU];
    }

    private get hasHeaderSlot(): boolean {
        return !!this.$slots.header;
    }

    private get hasDefaultSlot(): boolean {
        return !!this.$slots.default;
    }

    private get hasFooterSlot(): boolean {
        return !!this.$slots.footer;
    }

    private get hasPaddingHeader(): boolean {
        return this.paddingHeader && this.padding;
    }

    private get hasPaddingBody(): boolean {
        return this.paddingBody && this.padding;
    }

    private get hasPaddingFooter(): boolean {
        return this.paddingFooter && this.padding;
    }

    private onClick(): void {
        this.$emit('click');
    }
}

const PanelPlugin: PluginObject<any> = {
    install(v, options): void {
        v.prototype.$log.debug(PANEL_NAME, 'plugin.install');
        v.component(PANEL_NAME, MPanel);
    }
};

export default PanelPlugin;
