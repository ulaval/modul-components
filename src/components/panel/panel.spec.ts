import { createLocalVue, mount } from '@vue/test-utils';
import Vue, { VueConstructor } from 'vue';

import { renderComponent } from '../../../tests/helpers/render';
import { MPanel, MPanelSkin } from './panel';

describe('MPanel', () => {
    let localVue: VueConstructor<Vue>;
    beforeEach(() => {
        localVue = createLocalVue();
    });

    it('should render correctly', () => {
        const pnl = mount(MPanel, {
            localVue: localVue,
            slots: {
                default: 'body'
            }
        });

        return expect(renderComponent(pnl.vm)).resolves.toMatchSnapshot();
    });

    it('should render correctly when an header is provided', () => {
        const pnl = mount(MPanel, {
            localVue: localVue,
            slots: {
                default: 'body',
                header: 'header'
            }
        });

        return expect(renderComponent(pnl.vm)).resolves.toMatchSnapshot();
    });

    it('should render correctly when a footer is provided', () => {
        const pnl = mount(MPanel, {
            localVue: localVue,
            slots: {
                default: 'body',
                footer: 'footer'
            }
        });

        return expect(renderComponent(pnl.vm)).resolves.toMatchSnapshot();
    });

    it('should render skink correctly', async () => {
        const pnl = mount(MPanel, {
            localVue: localVue,
            slots: {
                default: 'body'
            }
        });

        for (const skin in MPanelSkin) {
            pnl.setProps({ skin: MPanelSkin[skin] });
            expect(renderComponent(pnl.vm)).resolves.toMatchSnapshot(skin);
        }
    });

    it('should render correctly with border', () => {
        const pnl = mount(MPanel, {
            localVue: localVue,
            propsData: {
                border: true
            }
        });

        return expect(renderComponent(pnl.vm)).resolves.toMatchSnapshot();
    });

    it('should render correctly with shadow', () => {
        const pnl = mount(MPanel, {
            localVue: localVue,
            propsData: {
                shadow: true
            }
        });

        return expect(renderComponent(pnl.vm)).resolves.toMatchSnapshot();
    });

    it('should render correctly with no padding', () => {
        const pnl = mount(MPanel, {
            localVue: localVue,
            propsData: {
                padding: false
            }
        });

        return expect(renderComponent(pnl.vm)).resolves.toMatchSnapshot();
    });

    it('should render correctly with no body padding', () => {
        const pnl = mount(MPanel, {
            localVue: localVue,
            propsData: {
                paddingBody: false
            },
            slots: {
                default: 'body'
            }
        });

        return expect(renderComponent(pnl.vm)).resolves.toMatchSnapshot();
    });

    it('should render correctly with no header padding', () => {
        const pnl = mount(MPanel, {
            localVue: localVue,
            propsData: {
                paddingHeader: false
            },
            slots: {
                header: 'header'
            }
        });

        return expect(renderComponent(pnl.vm)).resolves.toMatchSnapshot();
    });

    it('should render correctly with no footer padding', () => {
        const pnl = mount(MPanel, {
            localVue: localVue,
            propsData: {
                paddingFooter: false
            },
            slots: {
                footer: 'footer'
            }
        });

        return expect(renderComponent(pnl.vm)).resolves.toMatchSnapshot();
    });

    it('should emit click event when panel is clicked on', () => {
        const pnl = mount(MPanel, {
            localVue: localVue,
            slots: {
                default: 'body'
            }
        });

        pnl.find('article').trigger('click');

        expect(pnl.emitted('click'));
    });
});
