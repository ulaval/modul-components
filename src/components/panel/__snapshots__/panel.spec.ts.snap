// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MPanel should render correctly 1`] = `
<article class="m-panel m--is-skin-light m--has-border">
    <div class="m-panel__body m--has-padding-top m--has-padding-bottom">body</div>
</article>
`;

exports[`MPanel should render correctly when a footer is provided 1`] = `
<article class="m-panel m--is-skin-light m--has-border">
    <div class="m-panel__body m--has-padding-top">body</div>
    <footer class="m-panel__footer">footer</footer>
</article>
`;

exports[`MPanel should render correctly when an header is provided 1`] = `
<article class="m-panel m--is-skin-light m--has-border">
    <header class="m-panel__header">
        <div class="m-panel__header__content">header</div>
    </header>
    <div class="m-panel__body m--has-padding-bottom">body</div>
</article>
`;

exports[`MPanel should render correctly with border 1`] = `<article class="m-panel m--is-skin-light m--has-border"> </article>`;

exports[`MPanel should render correctly with no body padding 1`] = `
<article class="m-panel m--is-skin-light m--has-border">
    <div class="m-panel__body m--no-padding">body</div>
</article>
`;

exports[`MPanel should render correctly with no footer padding 1`] = `
<article class="m-panel m--is-skin-light m--has-border">
    <footer class="m-panel__footer m--no-padding">footer</footer>
</article>
`;

exports[`MPanel should render correctly with no header padding 1`] = `
<article class="m-panel m--is-skin-light m--has-border">
    <header class="m-panel__header m--no-padding">
        <div class="m-panel__header__content">header</div>
    </header>
</article>
`;

exports[`MPanel should render correctly with no padding 1`] = `<article class="m-panel m--is-skin-light m--has-border"> </article>`;

exports[`MPanel should render correctly with shadow 1`] = `<article class="m-panel m--is-skin-light m--has-shadow m--has-border"> </article>`;

exports[`MPanel should render skink correctly: Dark 1`] = `
<article class="m-panel m--is-skin-dark m--has-border">
    <div class="m-panel__body m--has-padding-top m--has-padding-bottom">body</div>
</article>
`;

exports[`MPanel should render skink correctly: Darker 1`] = `
<article class="m-panel m--is-skin-darker m--has-border">
    <div class="m-panel__body m--has-padding-top m--has-padding-bottom">body</div>
</article>
`;

exports[`MPanel should render skink correctly: Light 1`] = `
<article class="m-panel m--is-skin-light m--has-border">
    <div class="m-panel__body m--has-padding-top m--has-padding-bottom">body</div>
</article>
`;
